drivers/adc/adc_b91.c

```diff
@@ -0,0 +1,369 @@
+	}
+
+	status = adc_b91_validate_buffer_size(sequence);
+	if (status) {
+		LOG_ERR("Buffer size too small.");
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+/* Convert dts pin to B91 SDK pin */
+static adc_input_pin_def_e adc_b91_get_pin(uint8_t dt_pin)
+{
+	adc_input_pin_def_e adc_pin = NOINPUTN;
+
+	switch (dt_pin) {
+	FOR_EACH(CASE_DT_GPIO_TO_HAL_GPIO, (),
+		 ADC_GPIO_PB0, ADC_GPIO_PB1, ADC_GPIO_PB2, ADC_GPIO_PB3,
+		 ADC_GPIO_PB4, ADC_GPIO_PB5, ADC_GPIO_PB6, ADC_GPIO_PB7,
+		 ADC_GPIO_PD0, ADC_GPIO_PD1, ADC_VBAT)
+	}
+
+	return adc_pin;
+}
+
+/* Get ADC value */
+signed short adc_b91_get_code(void)
+{
+	signed short adc_code;
+
+	analog_write_reg8(areg_adc_data_sample_control,
+			  analog_read_reg8(areg_adc_data_sample_control) | FLD_NOT_SAMPLE_ADC_DATA);
+
+	adc_code = analog_read_reg16(areg_adc_misc_l);
+
+	analog_write_reg8(areg_adc_data_sample_control,
+		analog_read_reg8(areg_adc_data_sample_control) & (~FLD_NOT_SAMPLE_ADC_DATA));
+
+	if ((adc_code & BIT(SIGN_BIT_POSITION)) && !m_data.differential) {
+		adc_code = 0;
+	}
+
+	return adc_code;
+}
+
+/* ADC Context API implementation: start sampling */
+static void adc_context_start_sampling(struct adc_context *ctx)
+{
+	m_data.repeat_buffer = m_data.buffer;
+
+	adc_power_on();
+
+	k_sem_give(&m_data.acq_sem);
+}
+
+/* ADC Context API implementation: buffer pointer */
+static void adc_context_update_buffer_pointer(struct adc_context *ctx, bool repeat_sampling)
+{
+	if (repeat_sampling) {
+		m_data.buffer = m_data.repeat_buffer;
+	}
+}
+
+/* Start ADC measurements */
+static int adc_b91_adc_start_read(const struct device *dev, const struct adc_sequence *sequence)
+{
+	int status;
+
+	/* Validate input parameters */
+	status = adc_b91_validate_sequence(sequence);
+	if (status != 0) {
+		return status;
+	}
+
+	/* Set resolution */
+	switch (sequence->resolution) {
+	case 14:
+		adc_set_resolution(ADC_RES14);
+		m_data.resolution_offset = 0;
+		break;
+	case 12:
+		adc_set_resolution(ADC_RES12);
+		m_data.resolution_offset = 2;
+		break;
+	case 10:
+		adc_set_resolution(ADC_RES10);
+		m_data.resolution_offset = 4;
+		break;
+	case 8:
+		adc_set_resolution(ADC_RES8);
+		m_data.resolution_offset = 6;
+		break;
+	default:
+		LOG_ERR("Selected ADC resolution is not supported.");
+		return -EINVAL;
+	}
+
+	/* Save buffer */
+	m_data.buffer = sequence->buffer;
+
+	/* Start ADC conversion */
+	adc_context_start_read(&m_data.ctx, sequence);
+
+	return adc_context_wait_for_completion(&m_data.ctx);
+}
+
+/* Main ADC Acquisition thread */
+static void adc_b91_acquisition_thread(const struct device *dev)
+{
+	int16_t adc_code;
+
+	while (true) {
+		/* Wait for Acquisition semaphore */
+		k_sem_take(&m_data.acq_sem, K_FOREVER);
+
+		/* Wait for ADC data ready */
+		while ((analog_read_reg8(AREG_ADC_DATA_STATUS) & ADC_DATA_READY)
+				!= ADC_DATA_READY) {
+		}
+
+		/* Perform read */
+		adc_code = (adc_b91_get_code() >> m_data.resolution_offset);
+		adc_code = adc_code << (m_data.differential ? 0 : 1);
```

Comment:
`adc_code` is a signed integer, so when its value is negative, the result of `>>` is implementation-defined and the behavior of `<<` is undefined. I think you should rather use multiplication here.

Reply:
Done.